### PR TITLE
Yaw friction implementation

### PR DESCRIPTION
**Pull request by abhineet-gupta**
Not ready to merge: Testing 

**Feature or improvement description**
Add yaw friction model to OpenFAST based on code written by Rick Damiani in [his fork](https://github.com/rdamiani/openfast/tree/YawFriction)

**Impacted areas of the software**
Changes in elastodyn, new outputs added to OutlistParams.xlsx

Checklist:
- [x] Update regression tests
- [x] Add documentation to readthedocs
- [x] Verify readthedocs builds correctly
- [x] Add a few plots here to show how this works

### CODE REVIEW

**Comment by abhineet-gupta** [comment]
The yaw friction model plot is
image

**Comment by abhineet-gupta** [comment]
The changes to the documentation can be previewed here:
[YawFrictionInputFiles.pdf](https://github.com/OpenFAST/openfast/files/14302999/YawFrictionInputFiles.pdf)
[YawFrictionTheory.pdf](https://github.com/OpenFAST/openfast/files/14303001/YawFrictionTheory.pdf)

**Comment by andrew-platt** [comment]
A few quick comments

### Documentation section 4.8.3.4
 
Screenshot 2024-02-15 at 7 49 59 PM

See note at code location -- space needed before `:math:`.

### Documentation section 4.8.2.2.8
There is a discrepancy between the description of `YawFrctMod` and `M_CSmax`.
1. `YawFrctMod` indicates mode 1 has the `Fz` term, but `M_CSmax` indicates mode 2 has `Fz`.  I don't know which agrees with the code.
2. the description of `YawFrctMod` might need refinement.  Maybe something like `{0: none, 1: friction without Fz term at the yaw bearing, 2: friction includes Fz term at yaw bearing, 3: user defined model}` (or whatever the correct version is given 1.)
3. Descriptions in input files will need updating to reflect 1.

File: modules/elastodyn/src/ElastoDyn.f90 [hunk 1]
```diff
@@ -4440,229 +4463,10 @@ SUBROUTINE SetOutParam(OutList, p, ErrStat, ErrMsg )
 
 
 !   ..... Developer must add checking for invalid inputs here: .....
-if (p%BD4Blades) then
```
Comment (andrew-platt) [inline modules/elastodyn/src/ElastoDyn.f90:4443]: This block of code is missing.  We'll need to put it back in.

File: modules/elastodyn/src/ElastoDyn.f90 [hunk 2]
```diff
@@ -6591,6 +6396,61 @@ SUBROUTINE Teeter( t, p, TeetDef, TeetRate, TeetMom )
...
+
+      YawFriMf = 0.0_ReKi
+
+
+   CASE ( 1_IntKi, 2_IntKi )              ! 1= no Fz use. 2=Fz used
+
+      temp = -1.0_ReKi  !In the case of YawFrctMod=1 
+      
+      IF  (p%YawFrctMod .EQ. 2) THEN   
+        temp = MIN(0.0_R8Ki, Fz)  !In the case of YawFrctMod=1
```
Comment (andrew-platt) [inline modules/elastodyn/src/ElastoDyn.f90:6429]: comment `YawFrctMod=1` ?

File: modules/elastodyn/src/ElastoDyn.f90 [hunk 3]
```diff
@@ -6591,6 +6396,61 @@ SUBROUTINE Teeter( t, p, TeetDef, TeetRate, TeetMom )
...
 END SUBROUTINE Teeter
 !----------------------------------------------------------------------------------------------------------------------------------
+!> This routine computes the Yaw Friction Torque due to yaw rate and acceleration.
+SUBROUTINE YawFriction( t, p, Fz, Mzz, Omg, OmgDot, YawFriMf )
+!..................................................................................................................................
+
+      ! Passed Variables:
+   REAL(DbKi), INTENT(IN)             :: t                                       !< simulation time
+   TYPE(ED_ParameterType), INTENT(IN) :: p                                       !< parameters from the structural dynamics module
+   REAL(R8Ki), INTENT(IN )            :: Fz, Mzz                                 !< The teeter deflection, x%QT(DOF_Teet).
```
Comment (andrew-platt) [inline modules/elastodyn/src/ElastoDyn.f90:6406]: `teeter deflection`?  I think the note is wrong.

File: modules/elastodyn/src/ElastoDyn.f90 [hunk 4]
```diff
@@ -6591,6 +6396,61 @@ SUBROUTINE Teeter( t, p, TeetDef, TeetRate, TeetMom )
...
+   REAL(R8Ki), INTENT(IN )            :: Omg                                !< The yaw rate (rotational speed), x%QDT(DOF_Yaw).
+   REAL(R8Ki), INTENT(IN )            :: OmgDot                             !< The yaw acceleration (derivative of rotational speed), x%QD2T(DOF_Yaw).
+   
+   REAL(ReKi), INTENT(OUT)            :: YawFriMf                                 !< The total friction torque (Coulomb + viscous).
+
+      ! Local variables:
+   REAL(ReKi)                         :: temp                                   ! It takes teh value of Fz or -1.
+
+   
+   SELECT CASE ( p%YawFrctMod  ) ! Which friction model are we using? 0=None, 1=does not use Fz, 2=does use Fz
```
Comment (andrew-platt) [inline modules/elastodyn/src/ElastoDyn.f90:6416]: Compare with documentation and descriptions in the input files on `YawFrctMod`.

**Review by andrew-platt — CHANGES_REQUESTED** [review]
I didn't go through the `FixYawFric` routine in detail, but found a few other things that need correction.

**Comment by andrew-platt** [comment]
ToDo:

- [ ] add time series plots of yaw motion for `AOC_YawFriction_Loading` here
- [ ] add time series plots of yaw motion for `AOC_YawFriction_Stiffness` here

File: modules/elastodyn/src/ElastoDyn.f90 (hunk 1 above)
Comment (abhineet-gupta) [reply modules/elastodyn/src/ElastoDyn.f90:4443]: Resolved with commit "7303f6cc5"

File: modules/elastodyn/src/ElastoDyn.f90 (hunk 2 above)
Comment (abhineet-gupta) [reply modules/elastodyn/src/ElastoDyn.f90:6429]: Resolved with commit "7303f6cc5"

File: modules/elastodyn/src/ElastoDyn.f90 (hunk 4 above)
Comment (abhineet-gupta) [reply modules/elastodyn/src/ElastoDyn.f90:6416]: Compared and checked

**Comment by abhineet-gupta** [comment]
Yaw output for `AOC_YawFriction_Loading`:

AOC_YFriction_Loading-NacYaw
AOC_YFriction_Loading-NacYawV
AOC_YFriction_Loading-NacYawA
AOC_YFriction_Loading-YawFriMom


And for `AOC_YawFriction_Stiffness`:
NacYaw
NacYawV
NawYawA
YawFriMom

**Review by andrew-platt — APPROVED** [review]

**Comment by andrew-platt** [comment]
At final merge, we need to add the following credits:

- [ ] Co-authored-by:Kevin Fletcher <kbf4mc@virginia.edu>
- [ ] Co-authored-by:rdamiani <r.damiani@RRDengineering.com>

**Pull request merged**